mobile/src/main/java/org/openhab/habdroid/ui/WidgetAdapter.kt

```diff
@@ -557,6 +584,185 @@ class WidgetAdapter(
+                    showDatePicker(dt, widget, false)
+                } else if (widget.inputHint == Widget.InputTypeHint.Datetime) {
+                    showDatePicker(dt, widget, true)
+                } else if (widget.inputHint == Widget.InputTypeHint.Time) {
+                    showTimePicker(dt, widget, false)
+                }
+            }
+        }
+
+        override fun bind(widget: Widget) {
+            isBinding = true
+            updateJob?.cancel()
+
+            super.bind(widget)
+
+            val item = widget.item
+
+            inputText.inputType = when (widget.inputHint) {
+                Widget.InputTypeHint.Number -> TYPE_CLASS_NUMBER or TYPE_NUMBER_FLAG_DECIMAL or TYPE_NUMBER_FLAG_SIGNED
+                Widget.InputTypeHint.Date -> TYPE_CLASS_DATETIME or TYPE_DATETIME_VARIATION_DATE
+                Widget.InputTypeHint.Time -> TYPE_CLASS_DATETIME or TYPE_DATETIME_VARIATION_TIME
+                Widget.InputTypeHint.Datetime -> TYPE_CLASS_DATETIME or TYPE_DATETIME_VARIATION_NORMAL
+                else -> TYPE_CLASS_TEXT
+            }
+
+            val displayState = widget.stateFromLabel?.replace("\n", "") ?: ""
+
+            inputTextLayout.placeholderText = when {
+                widget.state != null -> ""
+                widget.inputHint == Widget.InputTypeHint.Text &&
+                    item?.isOfTypeOrGroupType(Item.Type.DateTime) == true -> "YYYY-MM-DD hh:mm:ss"
+                else -> displayState
+            }
+
+            val dataState = when {
+                widget.state == null -> ""
+                widget.inputHint == Widget.InputTypeHint.Number -> widget.state.asNumber?.formatValue()
+                displayState.isNotEmpty() -> displayState
+                widget.item?.isOfTypeOrGroupType(Item.Type.DateTime) == true -> when (widget.inputHint) {
+                    Widget.InputTypeHint.Date -> widget.state.asDateTime?.toISOLocalDate()
+                    Widget.InputTypeHint.Time -> widget.state.asDateTime?.toISOLocalTime()
+                    else -> widget.state.asDateTime?.toString()
+                }
+                else -> widget.state.toString()
+            }
+            inputText.setText(dataState)
+            oldValue = dataState
+
+            inputTextLayout.suffixText = when (widget.inputHint) {
+                Widget.InputTypeHint.Number -> widget.state?.asNumber?.unit
+                else -> null
+            }
+
+            // Don't directly edit field for date/time when inputHint set, but open popup when clicked
+            val isEditable = !(
+                    widget.inputHint == Widget.InputTypeHint.Date ||
+                    widget.inputHint == Widget.InputTypeHint.Time ||
+                    widget.inputHint == Widget.InputTypeHint.Datetime
+                )
+            inputText.isCursorVisible = isEditable
+            inputText.isFocusable = isEditable
+            inputText.isClickable = isEditable
```

Comment:
Won't this lead to the picker not shown anymore, since the `OnClickListener` isn't called if the view isn't clickable?
If the EditText needs to be clickable to be focusable, we should leave `isClickable` alone here and just set/unset the `OnClickListener`.

Reply:
It does show when Clickable is false. Unless this is something with the emulator, I think that is what it should be. I found discussions about it [here](https://stackoverflow.com/questions/14933330/datepicker-how-to-popup-datepicker-when-click-on-edittext).